src/dbnode/storage/index/block.go

```diff
@@ -162,32 +471,242 @@ func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
 	}
 
 	partialErr, ok := err.(*m3ninxindex.BatchPartialError)
-	if !ok { // should never happen
-		err := b.unknownWriteBatchInvariantError(err)
-		// NB: marking all the inserts as failure, cause we don't know which ones failed
+	if !ok {
+		// NB: marking all the inserts as failure, cause we don't know which ones failed.
 		inserts.MarkUnmarkedEntriesError(err)
 		return WriteBatchResult{NumError: int64(inserts.Len())}, err
 	}
 
 	numErr := len(partialErr.Errs())
 	for _, err := range partialErr.Errs() {
-		// Avoid marking these as success
+		// Avoid marking these as success.
 		inserts.MarkUnmarkedEntryError(err.Err, err.Idx)
 	}
 
-	// mark all non-error inserts success, so we don't repeatedly index them
+	// Mark all non-error inserts success, so we don't repeatedly index them.
 	inserts.MarkUnmarkedEntriesSuccess()
 	return WriteBatchResult{
 		NumSuccess: int64(inserts.Len() - numErr),
 		NumError:   int64(numErr),
 	}, partialErr
 }
 
-func (b *block) executorWithRLock() (search.Executor, error) {
-	var expectedReaders int
-	if b.activeSegment != nil {
-		expectedReaders++
+func (b *block) foregroundCompactWithBuilder(builder segment.DocumentsBuilder) error {
+	// We inserted some documents, need to compact immediately into a
+	// foreground segment.
+	b.Lock()
+	foregroundSegments := b.foregroundSegments
+	b.Unlock()
+
+	segs := make([]compaction.Segment, 0, len(foregroundSegments)+1)
+	segs = append(segs, compaction.Segment{
```

Comment:
intentional that the newest one comes first, or doesn't master?

Reply:
Doesn't really matter, the planner sorts everything first and does its own selection/grouping, etc regardless of input order.